jumanji/environments/routing/rware/viewer.py

```diff
@@ -0,0 +1,332 @@
+
+    def _draw_grid(self, ax: plt.Axes) -> None:
+        """Draw grid of warehouse floor."""
+        lines = []
+        # VERTICAL LINES
+        for r in range(self.rows + 1):
+            lines.append(
+                [
+                    (0, (self.grid_size + 1) * r + 1),
+                    ((self.grid_size + 1) * self.cols, (self.grid_size + 1) * r + 1),
+                ]
+            )
+
+        # HORIZONTAL LINES
+        for c in range(self.cols + 1):
+            lines.append(
+                [
+                    ((self.grid_size + 1) * c + 1, 0),
+                    ((self.grid_size + 1) * c + 1, (self.grid_size + 1) * self.rows),
+                ]
+            )
+
+        lc = LineCollection(lines, colors=(_GRID_COLOR,))
+        ax.add_collection(lc)
+
+    def _draw_goals(self, ax: plt.Axes) -> None:
+        """Draw goals, i.e. positions where shelves should be delivered."""
+        for goal in self.goals:
+            x, y = goal
+            y = self.rows - y - 1  # pyglet rendering is reversed
+            ax.fill(  # changed to ax, from plt, check if still works!
+                [
+                    x * (self.grid_size + 1) + 1,
+                    (x + 1) * (self.grid_size + 1),
+                    (x + 1) * (self.grid_size + 1),
+                    x * (self.grid_size + 1) + 1,
+                ],
+                [
+                    y * (self.grid_size + 1) + 1,
+                    y * (self.grid_size + 1) + 1,
+                    (y + 1) * (self.grid_size + 1),
+                    (y + 1) * (self.grid_size + 1),
+                ],
+                color=_GOAL_COLOR,
+                alpha=1,
+            )
+
+    def _draw_shelves(self, ax: plt.Axes, shelves: chex.Array) -> None:
+        """Draw shelves at their respective positions.
+
+        Args:
+            shelves: a pytree of Shelf type containing shelves information.
+        """
+        for shelf_id in range(self.n_shelves):
+            shelf = tree_slice(shelves, shelf_id)
+            y, x = shelf.position.x, shelf.position.y
+            y = self.rows - y - 1  # pyglet rendering is reversed
+            shelf_color = _SHELF_REQ_COLOR if shelf.is_requested else _SHELF_COLOR
+
+            x_points = [
+                (self.grid_size + 1) * x + _SHELF_PADDING + 1,
+                (self.grid_size + 1) * (x + 1) - _SHELF_PADDING,
+                (self.grid_size + 1) * (x + 1) - _SHELF_PADDING,
+                (self.grid_size + 1) * x + _SHELF_PADDING + 1,
+            ]
+
+            y_points = [
+                (self.grid_size + 1) * y + _SHELF_PADDING + 1,
+                (self.grid_size + 1) * y + _SHELF_PADDING + 1,
+                (self.grid_size + 1) * (y + 1) - _SHELF_PADDING,
+                (self.grid_size + 1) * (y + 1) - _SHELF_PADDING,
+            ]
+
+            ax.fill(x_points, y_points, color=shelf_color)
+
+    def _draw_agents(self, ax: plt.Axes, agents: chex.Array) -> None:
```

Comment:
we can break down this function to:

compute_agent_coordinates,
determine_agent_color,
draw_single_agent,
draw_agent_direction

_draw_agents would be simplified to a loop that iterates over all agents, calling these functions for each one.